File: Task1C/task1c/src/main/java/com/cps2004/Laptop.java
package com.cps2004;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

public class Laptop extends Electronics {
    private double screenSize;
    private String operatingSystem;
    private int ram;

    // Default constructor
    public Laptop(){
        super();
        this.screenSize = 0;
        this.operatingSystem = "";
        this.ram = 0;
    }

    // Parameterized constructor
    public Laptop(String name, int id, Supplier supplier, int price, int volume, String manufacturer, int warranty, double screenSize, String operatingSystem, int ram){
        super(name, id, supplier, price, volume, manufacturer, warranty);
        this.screenSize = screenSize;
        this.operatingSystem = operatingSystem;
        this.ram = ram;
    }

    // Getters and setters
    public double getScreenSize() { return screenSize; }
    public String getOperatingSystem() { return operatingSystem; }
    public int getRam() { return ram; }
    public void setScreenSize(double screenSize) { this.screenSize = screenSize; }
    public void setOperatingSystem(String operatingSystem) { this.operatingSystem = operatingSystem; }
    public void setRam(int ram) { this.ram = ram; }

    // Display product information
    public void display_product_info() {
        super.display_product_info();
        System.out.println("Product screen size: " + screenSize);
        System.out.println("Product operating system: " + operatingSystem);
        System.out.println("Product ram: " + ram);
    }

    @Override
    // Calculate discounted price
    public double calculate_discount(String month, int quantity){
        if ((month.equals("January") || month.equals("February") || month.equals("March")) && quantity >= 10) {
            return this.getPrice() * 0.1;
        } else {
            return this.getPrice();
        }
    }

    // Save product information to file
    public void save(ObjectNode newItem) {
        newItem.put("type", "laptop");
        super.save(newItem);
        newItem.put("screenSize", this.screenSize);
        newItem.put("operatingSystem", this.operatingSystem);
        newItem.put("ram", this.ram);
    }

    // Load product information from file
    public void load(JsonNode node) {
        super.load(node);
        screenSize = node.get("screenSize").asDouble();
        operatingSystem = node.get("operatingSystem").asText();
        ram = node.get("ram").asInt();
    }
}

File: Task3C/laptop.cpp
#include "laptop.h"

// Default constructor
Laptop::Laptop() {
    this->screenSize = 0;
    this->operatingSystem = "";
    this->ram = 0;
}

// Parameterised constructor
Laptop::Laptop(string name, int price, Supplier* supplier, int volume, int productID, string manufacturer, int warranty, double screenSize, string operatingSystem, int ram) : Electronics(name, price, supplier, volume, productID, manufacturer, warranty) {
    this->screenSize = screenSize;
    this->operatingSystem = operatingSystem;
    this->ram = ram;
}

// Getters and setters
double Laptop::getScreenSize() {
    return this->screenSize;
}

string Laptop::getOperatingSystem() {
    return this->operatingSystem;
}

int Laptop::getRam() {
    return this->ram;
}

void Laptop::setScreenSize(double screenSize) {
    this->screenSize = screenSize;
}

void Laptop::setOperatingSystem(string operatingSystem) {
    this->operatingSystem = operatingSystem;
}

void Laptop::setRam(int ram) {
    this->ram = ram;
}

// Display product information
void Laptop::display_product_info() {
    Electronics::display_product_info();
    cout << "Screen Size: " << this->screenSize << endl;
    cout << "Operating System: " << this->operatingSystem << endl;
    cout << "RAM: " << this->ram << endl;
}

// Calculate discount
double Laptop::calculate_discount(string month, int quantity) {
    if ((month == "January" || month == "February" || month == "March") && quantity >= 10) {
        return this->getPrice() * 0.1;
    } else {
        return this->getPrice();
    }
}

// Save product information
void Laptop::save(ostream& out) const {
    out << "laptop" << "\n";
    Electronics::save(out);
    out << screenSize << "," << operatingSystem << "," << ram << "\n";
}

// Load product information
void Laptop::load(istream& in) {
    Electronics::load(in);

    string line;
    getline(in, line, '\n');
    stringstream ss(line);

    string token;
    getline(ss, token, ',');
    screenSize = stod(token);

    getline(ss, token, ',');
    operatingSystem = token;

    getline(ss, token, ',');
    ram = stoi(token);
}
